llvm/lib/Transforms/Utils/LoopIterCountsToAssumptions.cpp

```diff
@@ -0,0 +1,215 @@
+//===-- LoopIterCountsToAssumptions.cpp - add Loop assumes -*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountsToAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/Dominators.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-metadata"
+
+using namespace llvm;
+namespace {
+/// check basic loop rotation conditions
+bool isLoopRotateCandidate(const Loop &L) {
+
+  SmallVector<BasicBlock *> LoopLatches;
+  L.getLoopLatches(LoopLatches);
+  SmallVector<BasicBlock *> ExitBlocks;
+  L.getExitBlocks(ExitBlocks);
+  BranchInst *BI = dyn_cast<BranchInst>(L.getHeader()->getTerminator());
+  return L.isLoopExiting(L.getHeader()) && (BI && BI->isConditional()) &&
+         (LoopLatches.size() == 1 && ExitBlocks.size() == 1);
+}
+
+/// get AddRec scalar evolution, if it can be found in Op
+const SCEVAddRecExpr *getAddRecSCEV(Value *Op, ScalarEvolution *SE) {
+  if (!SE->isSCEVable(Op->getType()))
+    return nullptr;
+
+  return dyn_cast<SCEVAddRecExpr>(SE->getSCEV(Op));
+}
+
+/// get the SCEV of the loop that describes how the loop IV is modified
+const SCEVAddRecExpr *getLoopCmpIVAddRecExpr(ICmpInst &LoopCmpInstr,
```

Comment:
nit: we could return an optional pair of SCEV and unsigned instead of by-val-parameter return.